### PR TITLE
tests: warn against reading into libvirt domxml during e2e tests

### PR DESCRIPTION
**Pull request by dankenigsberg**
End-to-end tests should be testing the KubeVirt API and behaviour. They should typically start a VM with specific options and then verify that the guest behaves as expected. E2e tests should *not* look into KubeVirt or Kubernetes implementation details. This is important in order to keep e2e short, few and limited to their level. Libvirt and qemu are very important to KubeVirt, but they sit below the KubeVirt API. Good e2e test should not look at them.

For example, tests verifying that the libvirt domxml was build as expected should be converted to unit test of domxml rendering function. A test that checks that a memory hotplug modified the domxml should be changed to a test that ensures that the additional memory is available inside the guest.

This PR deprecates two helper functions that break test levels so that we remember to avoid them.

```release-note
NONE
```

### CODE REVIEW

File: tests/utils.go [hunk 1]
```diff
@@ -84,6 +84,7 @@ func NewRandomReplicaSetFromVMI(vmi *v1.VirtualMachineInstance, replicas int32)
 	return rs
 }
 
+// Deprecated: GetRunningVirtualMachineInstanceDomainXML should not be used by end-to-end tests
```
Comment (jean-edouard) [inline tests/utils.go:87]: Verifying we generate proper libvirt XML should indeed happen in unit tests.
Verifying libvirt populated the right defaults and whatnot should happen in libvirt.
However, verifying we properly updated the XML after a hotplug for example, is something that I think belongs to KubeVirt e2e... WDYT?

Comment (akalenyu) [reply tests/utils.go:87]: We could ensure the hotplug actually happened (new disk available or so) instead of relying on the libvirt change
This may not be trivial in some cases but it would make this a more honest "end to end" test

Comment (dankenigsberg) [reply tests/utils.go:87]: + a unit test for the function that updates the XML during hotplug.

Comment (jean-edouard) [reply tests/utils.go:87]: > + a unit test for the function that updates the XML during hotplug.

That would be libvirt

Comment (dankenigsberg) [reply tests/utils.go:87]: > > \+ a unit test for the function that updates the XML during hotplug.
> 
> That would be libvirt

Ah, so no point for a kubevirt-side unit test, similar to

> Verifying libvirt populated the right defaults and whatnot should happen in libvirt.

Comment (xpivarc) [reply tests/utils.go:87]: I think we would like to have a integration tests with libvirt. For that I think it could be enough to have launcher running and the whole e2e env would not be needed. This is possible future improvement.

**Comment by dankenigsberg** [comment]
Rebased. @jean-edouard , would you consider approving this?

**Comment by jean-edouard** [comment]
> Rebased. @jean-edouard , would you consider approving this?

I still have concerns about this.
I do agree that the vast majority of the ~80 calls to `GetRunningVirtualMachineInstanceDomainXML` and `GetRunningVMIDomainSpec` should be moved to unit tests.
Not all however. An obvious exception is the reporter collecting the libvirt XML as an artifact to help debug failed tests.
Also, some paththrough tests verify that the right thing appeared in the XML, which I think is a good thing.
There are also tests that look at the XML after a migration to ensure things are sane. We do mess with the XML on migration, and testing that in unit tests is hard at best.

I get your point, and I'd love to see the number of calls go down from 80 to maybe a single digit number, but 0 seems unreasonable.
TL;DR: IMO deprecation is too strong, we need a `// Discouraged` :)

**Comment by xpivarc** [comment]
> > Rebased. @jean-edouard , would you consider approving this?
> 
> I still have concerns about this. I do agree that the vast majority of the ~80 calls to `GetRunningVirtualMachineInstanceDomainXML` and `GetRunningVMIDomainSpec` should be moved to unit tests. Not all however. An obvious exception is the reporter collecting the libvirt XML as an artifact to help debug failed tests. Also, some paththrough tests verify that the right thing appeared in the XML, which I think is a good thing. There are also tests that look at the XML after a migration to ensure things are sane. We do mess with the XML on migration, and testing that in unit tests is hard at best.
> 
> I get your point, and I'd love to see the number of calls go down from 80 to maybe a single digit number, but 0 seems unreasonable. TL;DR: IMO deprecation is too strong, we need a `// Discouraged` :)


@jean-edouard 
Deprecating this has an effect that IDEs will highlight this. Would you be more inclined if we capture the thought to reduce the usage and keep it per exception? With addition that we would like to find alternative how to test the integration. @vladikr maybe something you can help with.

File: tests/utils.go (hunk 1 above)
Comment (stu-gott) [reply tests/utils.go:62]: I concur with Jed. The domain XML of libvirt is their published API. Ensuring their API works is the domain of libvirt. Ensuring we honor the API is our domain.

Why are we conflating functional and integration tests?

Comment (stu-gott) [reply tests/utils.go:62]: To be clear, I'm not saying there's no value in an integration test. I'm suggesting this might not be where that gets done.

**Comment by stu-gott** [comment]
Thanks Jed. Language can be tricky. You're correct that the word "deprecated" has a specific meaning that formally announces the permanent retirement of a function or feature.

File: tests/utils.go (hunk 1 above)
Comment (akalenyu) [reply tests/utils.go:62]: > Ensuring their API works is the domain of libvirt

In an ideal world, we would not have to test that the API works, but seeing as we're the bridge between libvirt (or, others in the future) and container runtimes, I think we can't unfortunately escape having to test the end result, at least to some degree.

**Comment by xpivarc** [comment]
LGTM and I would even agree on the strong assertion of deprecating. I think e2e tests should not use it and once and if we have some more focused tests (be it integration tests or some where we only spin up required components) then we should move the asserts there...

@jean-edouard PTAL once you feel better

**Comment by jean-edouard** [comment]
> LGTM and I would even agree on the strong assertion of deprecating. I think e2e tests should not use it and once and if we have some more focused tests (be it integration tests or some where we only spin up required components) then we should move the asserts there...
> 
> @jean-edouard PTAL once you feel better

No matter how we look at this, there is at the very least one legitimate use of those functions (in the reporter), so by definition they are not deprecated. (as mentioned above, I think there are additional valid uses, but that's beyond the point)

Now we have to work with what's available, and if deprecation is the least bad solution we could find to discourage people from using those functions, then fine let's do that.
TL;DR: I don't approve but I don't feel strongly enough about it to disapprove either.

**Comment by xpivarc** [comment]
@dankenigsberg Do you want to rebase this?

**Comment by dankenigsberg** [comment]
> @dankenigsberg Do you want to rebase this?

Would that help me find an approver?

**Comment by xpivarc** [comment]
> > @dankenigsberg Do you want to rebase this?
> 
> Would that help me find an approver?

@dankenigsberg Yes. As I believe it will help to mitigate an anti-pattern of writing unit tests within functional tests.

**Comment by dankenigsberg** [comment]
Sorry I missed your comment for so long.

> libvirt is a well-tested component that we trust for VM configuration.

Correct. This is why KubeVirt should not test it again. It is enough if we pass the correct domxml to libvirt (which we should test in a unit test). We should not spend our limited resources testing that libvirt modifies the domxml as we expect.

> There are legitimate cases, such as testing the placement of devices on controllers or setting the affinity of threads or even virtual memory - that we cannot simply test from inside the guest OS nor from the host perspective - simply because what we see on the host requires a lot of context to compare to the input.

Would you point here to the good test that use these functions?

**Review by jean-edouard — COMMENTED** [review]
/approve

**Comment by dankenigsberg** [comment]
@vladikr would you agree with the current text?

Our liberal usage of these functions, instead of  logging into the guest (when possible) is harming our quality https://github.com/kubevirt/kubevirt/blob/main/tests/hotplug/memory.go#L133

**Review by 0xFelix — COMMENTED** [review]
> I concur. This function has a place in the reporter package.

I agree with @xpivarc, what do you think about deprecating the exported functions to discourage use and to encourage removal of them while also introducing non-exported functions private to the reporter tests?

This way we could deprecate and hopefully eventually remove these functions while replacing them with non-exported functions that can't easily be used elsewhere.

**Comment by 0xFelix** [comment]
@xpivarc @jean-edouard @vladikr Would you agree to replacing the exported functions with package private ones and indeed deprecate them and to gather the tests that really need them in such a package? Of course, this is more of a long-term effort.

**Comment by jean-edouard** [comment]
> @xpivarc @jean-edouard @vladikr Would you agree to replacing the exported functions with package private ones and indeed deprecate them and to gather the tests that really need them in such a package? Of course, this is more of a long-term effort.

Making multiple private copies of the function just so we can pretend the main one is deprecated? It doesn't make sense to me, but go for it if it makes people happy :)

**Comment by 0xFelix** [comment]
Making multiple private copies, that wasn't my idea. I hoped we could remove/replace it where possible and that only a few places would remain.
